File: front/src/shared/api/billing.ts
import type { Language, PlanId } from '@/shared/types/domain';

import { api } from './client';

export interface PlanPrice {
  language: Language;
  currency: string;
  /** Minor units — cents, kopecks — never a float. */
  amount: number;
}

export interface PlanOffer {
  plan: PlanId;
  prices: PlanPrice[];
  /** Null means no limit, and the client renders no counter for it. */
  activeTaskLimit: number | null;
  taskLinkLimit: number | null;
}

export interface Plans {
  /** False while no payment provider is wired: buying grants the plan outright. */
  billingEnabled: boolean;
  plans: PlanOffer[];
}

export interface Subscription {
  plan: PlanId;
  expiresAt: string | null;
  expired: boolean;
}

/** Public: the pricing page is one of the two that needs no account. */
export async function plans() {
  const answer = await api.get<{
    billingEnabled: boolean;
    plans: Array<Omit<PlanOffer, 'activeTaskLimit' | 'taskLinkLimit'> & {
      activeTaskLimit?: number | null;
      taskLinkLimit?: number | null;
    }>;
  }>('/plans');

  return {
    billingEnabled: answer.billingEnabled,
    plans: answer.plans.map((offer) => ({
      ...offer,
      activeTaskLimit: offer.activeTaskLimit ?? null,
      taskLinkLimit: offer.taskLinkLimit ?? null,
    })),
  } satisfies Plans;
}

/**
 * Buys a plan.
 *
 * The plan travels in the access token, so the account still reports the old
 * one until the token is rotated; the caller refreshes the session after this.
 */
export async function checkout() {
  const answer = await api.post<{ plan: PlanId; expiresAt?: string | null; expired: boolean }>(
    '/billing/checkout',
  );
  return { ...answer, expiresAt: answer.expiresAt ?? null } satisfies Subscription;
}

File: back/internal/adapter/http-handler/billing.go
package httphandler

import (
	"net/http"

	gen "github.com/moxicom/cursed_matrix/back/internal/adapter/http-handler/gen"
	"github.com/moxicom/cursed_matrix/back/internal/app/billing"
	"github.com/moxicom/cursed_matrix/back/internal/domain/shared"
	"github.com/moxicom/cursed_matrix/back/internal/domain/user"
)

type plansResponse struct {
	BillingEnabled bool            `json:"billingEnabled"`
	Plans          []gen.PlanOffer `json:"plans"`
}

// ListPlans answers with the pricing page, which needs no account.
func (a *API) ListPlans(w http.ResponseWriter, r *http.Request) {
	offers := a.billing.Offers()

	response := plansResponse{
		BillingEnabled: a.billing.Enabled(),
		Plans:          make([]gen.PlanOffer, 0, len(offers)),
	}
	for i := range offers {
		response.Plans = append(response.Plans, renderOffer(&offers[i]))
	}
	WriteJSON(w, r, http.StatusOK, response)
}

// Checkout buys a plan.
func (a *API) Checkout(w http.ResponseWriter, r *http.Request) {
	userID, ok := UserFrom(r.Context())
	if !ok {
		WriteError(w, r, shared.NewError(shared.CodeSessionExpired, nil))
		return
	}

	granted, err := a.billing.Checkout(r.Context(), userID)
	if err != nil {
		WriteError(w, r, err)
		return
	}
	WriteJSON(w, r, http.StatusOK, a.renderSubscription(granted))
}

func renderOffer(offer *billing.Offer) gen.PlanOffer {
	rendered := gen.PlanOffer{
		Plan:   gen.Plan(offer.Plan),
		Prices: make([]gen.PlanPrice, 0, len(offer.Prices)),
	}
	for _, price := range offer.Prices {
		rendered.Prices = append(rendered.Prices, gen.PlanPrice{
			Language: gen.Language(price.Language),
			Currency: price.Currency,
			Amount:   price.Amount,
		})
	}

	// Unlimited is the absence of a limit, not a number: the client renders no
	// counter for null, and a large integer would draw one.
	if offer.ActiveTaskLimit != user.Unlimited {
		limit := offer.ActiveTaskLimit
		rendered.ActiveTaskLimit = &limit
	}
	if offer.TaskLinkLimit != user.Unlimited {
		limit := offer.TaskLinkLimit
		rendered.TaskLinkLimit = &limit
	}
	return rendered
}

func (a *API) renderSubscription(subscription user.Subscription) gen.Subscription {
	return gen.Subscription{
		Plan:      gen.Plan(subscription.Plan),
		ExpiresAt: subscription.ExpiresAt,
		Expired:   subscription.Expired(a.clock.Now().UTC()),
	}
}
